File: leetcode/problems/py/permutations.py
class Solution(object):
    def permute(self, nums):
        returnArr = []

        if len(nums) == 1:
            return nums[:]
        
        for i in range(len(nums)):
            num = nums.pop(0)
            perms = self.permute(nums)

            for perm in perms:
                perm.append(num)
            returnArr.extend(perms)
            nums.append(num)
        return returnArr;

File: leetcode/problems/java/permutations.java
import java.util.List;
import java.util.ArrayList;

class Solution {
  public List<List<Integer>> permute(int[] nums) {
    List<List<Integer>> res = new ArrayList<>();
    backtrack(nums, 0, res);
    return res;
  }

  private void backtrack(int[] nums, int start, List<List<Integer>> res) {
    if (start == nums.length) {
      res.add(arrayToList(nums));
      return;
    }

    for (int i = start; i < nums.length; i++) {
      swap(nums, start, i);
      backtrack(nums, start + 1, res);
      swap(nums, start, i);
    }
  }

  private List<Integer> arrayToList(int[] arr) {
    List<Integer> list = new ArrayList<>();
    for (int num : arr) {
      list.add(num);
    }
    return list;
  }

  private void swap(int[] nums, int i, int j) {
    int temp = nums[i];
    nums[i] = nums[j];
    nums[j] = temp;
  }
}